### DISCUSSION TITLE
Welcome to kago Discussions!

### DISCUSSION THREAD

**Discussion by kamalshkeir** [kamalshkeir/kago#1] [Ideas]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with anything in your mind.

**Comment by manson** [comment]
Great project! It is very similar to https://pocketbase.io/ project and I like both but every has its nuances. I wont tell you about pocketbase, you may interest it of your own, just tell about the only two things I wish to be changed here to be a perfect engine for small and middle-sized projects: 1) make better admin panel (see pocketbase) and a bit more automated migration. I dont want to think about it, I want to forget the migrations. And you almost have done this for me :-) If you would suggest fully automatic migration (check the model exists, if not - Automigrate, if yes - Link it) it would be perfect. Of cause if I need I will make an additional steps but this rather be a rear case... If possible, it would be great to make an additional interface to use document database (mongodb etc) or key-value storage with the same or similar usage... There are many more things to come up with this (go scripting for stored procedures-like functionality, RPC for this SPs, remote backup-restore (see total.js engine all-in-one for js), distributed funcs etc) just imagine :-)

**Reply by kamalshkeir** [reply]
So glad that you have see the potential behind KaGo, and thank you for this feedback, everything you said make sense for me, i will try to explain in details each point:

1- better admin panel: pocketbase have a very attractive admin panel, I see it on reddit few weeks ago, and planning to make a similar dashboard design and adding more functionalities, now, can i do it using go template engine ? while keeping the developer on control for all templates and static files cloned to the project ?, sound a bit complicated beside my everyday job without the help of the community, but i have another idea that you may like, what if i make the admin theme DYNAMIC, by default you will have the current theme, anyone could create a theme, put it on a github repo, and give the repo path to the app to be used as theme
P.S: there are a lot of admin features that you can easily miss, and not documented, for example: 
- pagination infinite scroll using IntersectionObserver
- Progressive Web App, you can install it on your phone
- when you delete a model with image, the image file is deleted automatically from your local storage 

2- for migrations, it's a little more complicated, because Go stay a compiled language that need to know types on compilation time, for now all my benchmarks against Gorm are better, because i use the minimum amount of reflection, and was not possible before generics.So on compilation time, i can check if every table in the database, have a model linked to it, if not , i didn't find a way to get type Param from code automaticaly without a orm.LinkModel[User](table_name), because you cannot passe type param dynamically, so i really think that it could not be easier without loosing performance.
btw ,  [golobby-orm](https://github.com/golobby/orm)  take different approach and still showing the limitation of generics in go for know, you should in addition create a method for each struct,

3- for document database, i have a working mongodb package that can be easily add it to the orm, i will try to add it this weekend

Remote backup restore is indeed the best next feature to implement, what do you think about sending backups per email too ?

I didn't know about total.js, i will check it  :)

**Reply by manson** [reply]
Hi! So many new good news! :-)

1) not quite sure about theming - in my point of view theming is all just about colors. but if you mean the admin UI itself - this is a good idea, but I cannot imagine how it would me implemented - you know - communication between UI and backend etc. Besides you mentioned go templates but this is a bit different - it is compiled (or server-side rendered) but other UIs may be total separate - like Vue, React etc and only communicates with backend via API. I prefer vue and it seems better for users to make their own,  but for admin panel it seems to be better go templates because admin panel is being used rear or by one user and initially will have more features itself. I think it would be good choice to use this framework (https://github.com/GoAdminGroup/go-admin) - it is rich, go-templated and designed to be included into the go apps as admin UI or just user UI. Besides, I think this may be dynamic as you say, although separate UI package like vue-based or react-based might be totally dynamic...

" there are a lot of admin features that you can easily miss" - sure I missed this :-) I just read info on front page of repository and tried to create a small app to "taste" this. And I liked it :-) And this features you specified are very good!

2) Migrations. Not sure what the difficulties you talk about. All the migrations (auto or not) are being run at startup time and possible long time running is not a big deal. Besides it's not good to perform something database-specific at compilation time because at run-time I can switch the database or just copy binary to an environment with a very old db structure. 
So, not quite clear this moment. Every property of model has it's description and this is read at runtime. So at any moment I can read all them and you do this ("orm:...") the same i can read columns of a table. Not sure what you do by LinkModel[User] function (will take a look later) but all I see you already have. The only thing I have to do ether run Migrate function or Link, right? Why cannot you check the table existence during LinkModel and run Migrate if table not exist? Of cause there could be complex migration algorithms, but this is user's responsibility. All that I as user should do at startup - something like LinkModel or Migrate - all the params you already have. Or did I miss something?

3) I love you! :-) Mongodb is my commonly used db for last years, so this is very and very well to hear this. The only thing to note - if you are about to supply to the user the standard mongo driver interface, I would suggest to use qmgo instead (https://github.com/qiniu/qmgo) It is based on standard driver but much, much easier to use, but if this is only the standard (your's) orm to be available, its ok. Anyway, take a look at qmgo, probably it will be useful for you in any case...

" what do you think about sending backups per email too ?" - its good for small projects and if user doesn't have other servers available, but I would prefer some other mechanizm like ftp or rsync  or sc  upload/copy or dropbox/google drive/ yandex disk integrations. Or even pure bittorent or bittorent-based syncs... Or the most simple thing - backup, zip and send to the owner not the file itself, but link to file that is served by application itself (not sure to send big backups is a good idea). Or even just an admin UI to access all the backups from UI with or without email notification. And of cause encryption of file if needed.

You may take a look at ntfy (https://ntfy.sh/) project or gotify (https://gotify.net/) projects to extend the notifications possibilities of your app...

"I didn't know about total.js, i will check it :)" - this is not about framework itself, just about idea how one framework can become a real combine in a good sense

**Reply by manson** [reply]
Just to clarify about migration. 
Stating: If table does not exists you have to run AutoMigrate else LinkModel
Problem: I don't know if this table exists. What should I do? AutoMigrate and catch error and if error rises run LinkModel? If yes, what prevent you doing this in one call Migrate? All the params (model, table name) already are being passed here. in this case I dont care if table exists or not, just run Migrate on every application run.

**Reply by kamalshkeir** [reply]
I was just waiting the break to be able to answer you lol.
i will answer first for migrations, you will never have a problem with AutoMigrate and LinkModel, you only use one of them depending on the case( if table exist already or your need to migrate a new one), and AutoMigrate  will never error if the table already exist in the database because of CREATE IF NOT EXIST.
For example if you need a more complicated migrations  you can do it from the shell using an SQL file , or if  you just connect to a new database with tables already inside, so in this case , the application will know about the new table created , but how can i find the model that you want to connect to this table without orm.LinkModel in this case?
So if you want to migrate a new table, do the AutoMigrate and keep it in your code, if table exist u create your struct for the table and link it using orm.LinkModel.

**Reply by kamalshkeir** [reply]
I really like this app https://ntfy.sh/ , very handy, AND very easy to make a clone of it using KaGo :=), i assume they are using Websockets + PWA service worker to push notifications, which is both are up and ready for you

**Reply by manson** [reply]
> 

So the common case for Post structs (ether for the first time for Post creation and when fields changed and just every start application) is to:
orm.AutoMigrate[Post]("database.sqlite", "posts", true)
orm.LinkModel\[Post ]("posts")

i.e. Just call both this on app startup and forget the headache? (simple cases - just adding, deleting columns)?

**Reply by kamalshkeir** [reply]
for the admin, i have make a small change yesterday: when you run the app first time, it will clone my assets repos where you find all templates and static files for admin, it is at /kamalshkeir/kago-assets 
Now if you want to create a new theme completely, not just styling, you have to do 3 simple steps, imagine you are a company you don't like this admin panel or others and want to create your own with your logo, your organisations of data, etc, so those 3 steps:
1 - you can modify all admin handlers to send you different data, like for admin login if you want to implement a two factor auth or JWT, and you have access directly to all html templates and static file to erase completely or to modify everything, btw all middlewares are also editable the same way for the same reason.
2 -  keep at least these 2 folders assets/static and assets/templates
3- upload the assets folder into a repo and set these 2 values:
settings.REPO_USER="YOUR_REPO_USER" // default kamalshkeir
settings.REPO_NAME="YOUR_REPO_NAME" // default kago-assets

**Reply by kamalshkeir** [reply]
> 
no never put both of them for the same table, you have only 2 cases:
1- if table exist already in db use only orm.LinkModel[Post ]("posts") // you keep it 
2- table not exist  use only orm.AutoMigrate[Post]("database.sqlite", "posts") // you keep it also
no need to remove automigrate after migrate, if table exist it will neither update the table, neither return an error, it will just link the model to the table

**Reply by manson** [reply]
" if table exist already in db" - you suggest me every run check if table exist? I may have two copies of apps - one stage (for separate checking changes) and one or several products. So I may have staged server updated sequentially but product - only the final commit, skipping intermediate changes. And headache with structures returns. As I say I want for simple cases (add table, column, remove column) just don't think on migrations - just put "migrate" with models on app startup and it check of its own if needs automigrate or just link model and table.  The landmark - working with document database where there are no all this "ado about nothing"...

**Reply by manson** [reply]
> for the admin  

Good changes

**Reply by kamalshkeir** [reply]
Now i get your point. you are right, it will be better to remove LinkModel and keep only AutoMigrate that handle everything, synchronise between table and model, maybe showing warning message for extra or missing fields, if exist in db it will check if any difference between the 2 versions, if any will update the database and link the model. I'm gonna have fun this weekend :)

**Reply by kamalshkeir** [reply]
are you ready ?, it was the best idea you had, it's so fuckin convinient , thank you ❤️ , i will documented and push it, it work like charm, try it and tell me what u think

**Reply by kamalshkeir** [reply]
lol, where are you @manson

**Reply by manson** [reply]
Hey! Congrats! Really like you did it! But I couldn't check it the weekend (had a lot of home jobs out of comp). Will try it today or tomorrow and will give you feedback :-) You're good, man!

**Reply by kamalshkeir** [reply]
Thank you :smile:, this feature let me awake all the night

**Reply by manson** [reply]
Hi!
I have played a bit with migrations. Itself migration works well, although I used some simple examples. But the question is in different. Its not quite clear how to work with databases from init - the docs lack this, I spent some time to make the right code sequence. Its better to have full example.   
Besides the default database name is used from .env file (admin and maybe some other parts), so if I want to use the default database in code I need to know its name from code anyway (read .env or remember what I used). Because database name is mandatory for NewDatabaseFromDSN that is used as default entrypoint to orm as I understand. Why not use default name if dbName is empty? I used code like this:
```
var dbName = "db"

func main() {	
	println("creating database ", dbName)
	if err := orm.NewDatabaseFromDSN("sqlite", dbName) ; err !=nil {
		panic(err)
	}		
	if err := orm.AutoMigrate[Bookmark]("bookmarks", dbName); err !=nil {
		panic(err)
	}
	app := kago.New()	
	app.Run()
}
```

Here I put the dbName in code but should read it from env, but it's better to use it as I wrote upper

The second small issue (not issue but good coding) - use enums for database types or constants, say orm.SQLITE instead of "sqlite" to be used in params.

Sorry, had no more time to look at this, will do it later

**Reply by kamalshkeir** [reply]
did you try to remove field from struct after migration ?

**Reply by manson** [reply]
yes, it removed it in db as well. So it looks great right now!

**Reply by kamalshkeir** [reply]
1- kago.New(".env") take an optional parameter for env file, because it load env and default everything, in your code here you kago.New() is not at the top of main, i didn't understand why you do it like this
2- all env vars are loaded to settings.Config, you can see it in the readme envloader part, so you can access dbName, dbType anytime using settings.Config.Db.Name,...
3- beside that u have orm.GetDatabases(), orm.GetDatabase(dbName) and orm.GetConnection(dbName) all defaulted to settings.Config.Db.Name, and dbName can be "default" to get the default connection or default db
4- i just added SQLITE,POSTGRES and MYSQL to orm package as constant

**Reply by kamalshkeir** [reply]
As @DanielcoderX said, i need at least a website for documentation

**Reply by kamalshkeir** [reply]
> yes, it removed it in db as well. So it looks great right now!

great, missing only the index and multi index constraint and we are good

**Reply by manson** [reply]
1 - I wanted to see if I can use orm standalone if I need to use only db-specific functionality. It looks good
2 - now I see this. I didn't go into deep, so missed some.
3 - I would suggest to add some real default values for, say, orm.Getdatabase() without dbName. If empty - it takes from config by default. That's because most time developer will deal with 1 database and it's name is set in .env. So dev dont want to make additional steps to get current database object from settings. Just orm.GetDatabase() and that's it. If he wants many databases it is his responsibility how he gets their objects (like it is now)
4 - nice! :-)

As for documentation - you can just fill the wiki part of github. It's enough at this step I think. The better is to move project to a stable stage.

**Reply by manson** [reply]
1 - I would rather split app parts more deeply to make this parts mostly independent. As I say, I didnt go deep in code, but it seems that kago.New() initializes all the parts (orm as well). What if I dont want to use orm in this microservice? Or on the contrary want to write an utility that makes some dataprocessing in database only. So I would use "settings" object separately and pass it to app parts initialization I want. Yes, this is going to be more strings at the beginning of main procedure, but... kago.New() may init the whole app. But in other cases something like:
```
settings: = kago.LoadEnvironment()
orm.Init(settings) // or Init(settings.database)
router.Init(settings)
cache.Init(settings) // or cache.Init(orm) or cache.Init(router) whatever it needs...
whatever.Init(settings)
```

I do not insist on this. This is just for reflection... :-) Maybe you find some useful thoughts

**Reply by kamalshkeir** [reply]
cool, i will present it to my company soon, they seem's to like it, but like it not mean use it :p 

to answer your suggestion for orm.Getdatabase() i will make dbName optional parameter, make sense, but not remove it completely, because it take me forever to handle multiple database, and it's convenient, like sql builders they have methods Database to choose from which database you want to query, and in addition to that, there's 4 ways to get default database at the current time:
1- orm.Getdatabases()[0]
2- orm.Getdatabase(orm.DefaultDB)
3- orm.Getdatabase("")
4- orm.Getdatabase("default")

**Reply by manson** [reply]
"cool, i will present it to my company soon, they seem's to like it, but like it not mean use it :p" - I wish you good luck with this! :-)

" but not remove it completely" - sure not to remove! Multiple databases is great. For example to split database to several (say... for tenants or other logic) or to sync data from local database to remote etc.

**Reply by kamalshkeir** [reply]
AutoMigrate is now ready, it can handle foreign keys on delete and on cascade, indexes working now

**Comment by DanielcoderX** [comment]
Hi man :)
Perfect project 👌 
Doesn't want to make a web page for an introduction?

**Comment by charkops** [comment]
Hey, nice job on the project.

I was looking to contribute, but after a while I realized that the *.html files of the front facing admin page are not included in the commited codebase but instead managed separately ?
Is there a reason for this ? Wouldn't it be much easier for people to handle front-facing bugs if the files were included in this repo instead of a separate one ?

Cheers.

**Reply by kamalshkeir** [reply]
Hi @charkops, Thank you :) 

There is many reason why it is separated:
- Because it is cloned using git so if your static and templates files are embeded , static and templates will not be generated
- Anyone can create a new admin dashboard, or extend the default one, and put it on a custom repo, then change these values to use it instead of the default settings.REPO_USER="YOUR_REPO_USER" // default kamalshkeir
settings.REPO_NAME="YOUR_REPO_NAME" // default kago-assets
- files will be cloned the first time you create the project, or whenever assets folder is missing, so you have everything in your project

https://github.com/kamalshkeir/kago#generated-admin-dashboard

**Comment by kamalshkeir** [comment]
https://github.com/kamalshkeir/kago/discussions/7

**Reply by kamalshkeir** [reply]
can i have your advice guys on this bus ?, people on reddit are saying i am reinventing the wheel, but i could'nt find anything close to it so far

@charkops  @manson  @DanielcoderX

**Reply by DanielcoderX** [reply]
Hey man.
Maybe a wheel but not **True**.

This is perfect and I couldn't find sth like this(there was some similar but not as perfect as this)

It's not matter that people say you are making wheel
Sometimes You may need a wheel with different _stripe_ :)

**Reply by kamalshkeir** [reply]
lol, thank you :) , i was loosing motivation, really https://www.reddit.com/r/golang/comments/ycg06d/i_am_making_a_shared_event_bus_between_server_and/

btw , i have created a separated package github.com/kamalshkeir/kmux , only the router, and i will continue doing that for the others package, people will be more free to compose their application

**Reply by DanielcoderX** [reply]
Perfect :))
Good luck🙌🏻

**Reply by kamalshkeir** [reply]
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <h1>Index</h1>
    <button id="btn">Click</button>


<script src="/assets/js/Bus.js"></script>
<script>
let bus = new Bus("localhost:9313");

bus.OnOpen = (e) => {
    bus.Subscribe("client-js",(data) => {
        console.log("new=",data);
    })  
}

btn.addEventListener("click",(e) => {
    e.preventDefault();
    bus.SendTo("server:go","hello from index.html")
})

</script>
</body>
</html>
```

**Comment by kamalshkeir** [comment]
@charkops @manson @DanielcoderX go get it guys :) ,  i tried to publish it as fast as i can, still need a lot of tests, so you can help me with that, hope you will enjoy using it

**Reply by kamalshkeir** [reply]
@charkops @manson @DanielcoderX https://github.com/kamalshkeir/kbus

**Reply by DanielcoderX** [reply]
Hey man
 Checked it 
Perfect documentation :)
Sorry that I can't help you to pass the test (at the **exam** season) :(💔

**Comment by kamalshkeir** [comment]
@DanielcoderX  @manson @charkops  , yet another pkg i know ,lol, https://github.com/kamalshkeir/korm , and a tiny surprise for @manson 😂

**Reply by DanielcoderX** [reply]
Good for manson😂🙌🏻

Just can say it's going to be an ultra **perfect**  ORM ever that had been
Specially in **performance** bench's

**Reply by kamalshkeir** [reply]
lol @manson wanted so badly the mongodb

> Just can say it's going to be an ultra perfect ORM ever that had been . Specially in performance bench's

Thank you <3
